data_steward/cdr_cleaner/reporter.py

```diff
@@ -0,0 +1,306 @@
+        rule_info = {}
+
+        try:
+            clazz = rule[0]
+            instance = clazz('foo', 'bar', 'baz')
+            LOGGER.info(f"{clazz} ducktyped to a class")
+
+        except (RuntimeError, TypeError, HttpError, BadRequest):
+            LOGGER.info(f"{rule} did NOT ducktype to a class")
+            rule_info = get_function_info(clazz, fields_list)
+            report_rows.append(rule_info)
+
+        else:
+            try:
+                # this is a class
+                _ = instance.get_query_specs()
+                LOGGER.info(f"{clazz} is a class")
+                for field in fields_list:
+                    try:
+                        value = 'NO DATA'
+                        if field in report.FIELDS_PROPERTIES_MAP:
+                            func = report.FIELDS_PROPERTIES_MAP[field]
+                            value = getattr(instance, func, 'no data')
+                        elif field in report.FIELDS_METHODS_MAP:
+                            func = report.FIELDS_METHODS_MAP[field]
+                            value = getattr(instance, func, 'no data')()
+                        elif field in report.CLASS_ATTRIBUTES_MAP:
+                            func = report.CLASS_ATTRIBUTES_MAP[field]
+
+                            value = None
+                            for item in func.split('.'):
+                                if not value:
+                                    value = getattr(instance, item)
+                                else:
+                                    value = getattr(value, item)
+
+                        rule_info[field] = value
+
+                    except AttributeError:
+                        # an error occurred trying to access an expected attribute.
+                        # did the base class definition change recently?
+                        LOGGER.exception(
+                            f'An error occurred trying to get the value for {field}'
+                        )
+                        rule_info[field] = 'unknown'
+            except (TypeError, AttributeError):
+                # an error occurred indicating this is not a rule extending the
+                # base cleaning rule.  provide the info we can and move on.
+                LOGGER.exception(f'{clazz} is not a class')
+                # this is a function
+                rule_info = get_function_info(clazz, fields_list)
+
+            report_rows.append(rule_info)
+
+    return report_rows
+
+
+def separate_sql_statements(rules_values):
+    """
+    Separate SQL statements into items with other identical fields.
+
+    This must maintain the SQL statement order.
+    """
+    separated_rules_values = []
+    for rule_values in rules_values:
+        sql_list = []
+        separated = dict()
+
+        # gather the queries as a list
+        for query_dict in rule_values.get('sql', []):
+            sql_list.append(query_dict.get('query', ''))
```

Comment:
When a function returns zero queries the following error is returned:

```
Traceback (most recent call last):
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 2073, in <module>
    main()
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 2067, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 1418, in run
    return self._exec(is_module, entry_point_fn, module_name, file, globals, locals)
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 1425, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/aimeemcgrenera/Documents/dev/curation/data_steward/cdr_cleaner/reporter.py", line 306, in <module>
    main()
  File "/Users/aimeemcgrenera/Documents/dev/curation/data_steward/cdr_cleaner/reporter.py", line 299, in main
    write_csv_report(args.output_filepath, args.data_stage, args.fields)
  File "/Users/aimeemcgrenera/Documents/dev/curation/data_steward/cdr_cleaner/reporter.py", line 259, in write_csv_report
    required_fields_dict = format_values(required_fields_dict, fields_list)
  File "/Users/aimeemcgrenera/Documents/dev/curation/data_steward/cdr_cleaner/reporter.py", line 193, in format_values
    rules_values = separate_sql_statements(rules_values)
  File "/Users/aimeemcgrenera/Documents/dev/curation/data_steward/cdr_cleaner/reporter.py", line 163, in separate_sql_statements
    sql_list.append(query_dict.get('query', ''))
AttributeError: 'str' object has no attribute 'get'
```

Reply:
This should be accounted for now.  Thank you for catching this.